port/linux/omrvmem.c

```diff
@@ -1132,56 +1132,89 @@ reserve_memory_with_mmap(struct OMRPortLibrary *portLibrary, void *address, uint
 		flags |= MAP_SHARED;
 		useBackingSharedFile = TRUE;
 	} else {
 		useBackingFile = set_flags_for_mmap(&flags);
 	}
 
-	if (0 != (mode & OMRPORT_VMEM_MEMORY_MODE_MMAP_HUGE_PAGES)) {
+	if (OMR_ARE_ANY_BITS_SET(mode, OMRPORT_VMEM_MEMORY_MODE_MMAP_HUGE_PAGES)) {
 		flags |= MAP_HUGETLB;
 		areHugePagesEnabled = TRUE;
 	}
 
 	if (useBackingSharedFile) {
-		memfd_function_t memfd_createDL = PPG_memfd_function;
-		if (NULL != memfd_createDL) {
-			Trc_PRT_vmem_reserve_doubleMapAPI_available(address, byteAmount);
-			mode |= OMRPORT_VMEM_MEMORY_MODE_DOUBLE_MAP_AVAILABLE;
-			int ft = -1;
-			uintptr_t fdFlags = 0;
+		if (OMR_ARE_ANY_BITS_SET(mode, OMRPORT_VMEM_MEMORY_MODE_SHARE_TMP_FILE_OPEN)) {
+			/* Generate a unique temporary filename from template and open the file. */
 			char filename[FILE_NAME_SIZE + 1];
-			sprintf(filename, "omrvmem_temp_%zu_%09d_%zu",  (size_t)omrtime_current_time_millis(portLibrary),
-								getpid(),
-								(size_t)pthread_self());
-
-			if (areHugePagesEnabled) {
-				fdFlags = MFD_HUGETLB;
+			snprintf(filename, sizeof(filename), "omrvmem_%09d_XXXXXX", getpid());
+			fd = mkostemp(filename, 0);
+			if (OMRPORT_INVALID_FD != fd) {
+				unlink(filename);
+				/* Set the file size with ftruncate. */
+				if (OMRPORT_INVALID_FD == ftruncate(fd, byteAmount)) {
+					close(fd);
+					fd = OMRPORT_INVALID_FD;
+				}
 			}
-			fd = memfd_createDL(filename, fdFlags);
-			ft = ftruncate(fd, byteAmount);
-
-			if ((OMRPORT_INVALID_FD == fd) || (OMRPORT_INVALID_FD == ft)) {
-				Trc_PRT_vmem_reserve_failed(address, byteAmount);
-				update_vmemIdentifier(identifier, NULL, NULL, 0, 0, 0, 0, 0, NULL, -1);
-				Trc_PRT_vmem_reserve_exit(result, address, byteAmount);
-				return result;
+			if (OMRPORT_INVALID_FD == fd) {
+				Trc_PRT_vmem_reserve_tempfile_not_created(filename, byteAmount);
 			}
 		} else {
+			memfd_function_t memfd_createDL = PPG_memfd_function;
+			if (NULL != memfd_createDL) {
+				Trc_PRT_vmem_reserve_doubleMapAPI_available(address, byteAmount);
+				mode |= OMRPORT_VMEM_MEMORY_MODE_DOUBLE_MAP_AVAILABLE;
+				int ft = -1;
+				uintptr_t fdFlags = 0;
+				char filename[FILE_NAME_SIZE + 1];
+				sprintf(filename, "omrvmem_temp_%zu_%09d_%zu",  (size_t)omrtime_current_time_millis(portLibrary),
+									getpid(),
+									(size_t)pthread_self());
+
+				if (areHugePagesEnabled) {
+					fdFlags = MFD_HUGETLB;
+				}
+				fd = memfd_createDL(filename, fdFlags);
+				ft = ftruncate(fd, byteAmount);
+
+				if ((OMRPORT_INVALID_FD == fd) || (OMRPORT_INVALID_FD == ft)) {
+					Trc_PRT_vmem_reserve_failed(address, byteAmount);
+					update_vmemIdentifier(identifier, NULL, NULL, 0, 0, 0, 0, 0, NULL, -1);
+					Trc_PRT_vmem_reserve_exit(result, address, byteAmount);
+					return result;
+				}
+			} else {
+				Trc_PRT_vmem_reserve_doubleMapAPI_not_available(address, byteAmount);
+			}
+		}
+
+		if (OMRPORT_INVALID_FD == fd) {
+			/* Revert to using an ANONYMOUS mmap. */
+			useBackingSharedFile = FALSE;
```

Comment:
Please explain why doing something other than what the caller requested makes sense.